### PR TITLE
Controller class upgrade

### PR DESCRIPTION
**Pull request by olkorhon**
# Status
**FINISHED**

# Description
This PR refactors controllers to use Class architecture.

# Todos
### controllers
- [x] builds
- [x] campaigns
- [x] groups
- [x] index
- [x] items
- [x] loans
- [x] resources
- [x] results
- [x] targets
- [x] testcases
- [x] users

### cleanup
- [x] Unittests
- [ ] ~Functional testing~
- [ ] ~Documentation~

# Impacted areas
* controllers
* routes

### CODE REVIEW

**Comment by olkorhon** [comment]
No labels for in progress pull requests in this repository, maybe something like "do not merge" would be useful. @jupe

File: app/controllers/index.js [hunk 1]
```diff
@@ -69,16 +57,17 @@ class DefaultController extends EventEmitter {
...
+      const errorMsg = `get failed: Cannot get model, request does not have a value linked to key: ${this.modelName}`;
       winston.warn(errorMsg);
-      res.status(300).json({ error: errorMsg });
+      res.status(500).json({ error: errorMsg });
     }
   }
 
   find(req, res) {
-    this._model.query(req.query, (error, list) => {
+    this._model.find(req.query, (error, list) => {
```
Comment (jupe) [inline app/controllers/index.js]: why this change? all models are extended with `mongoose-query` -plugin (like [this](https://github.com/OpenTMI/opentmi/blob/master/app/models/resource.js#L178)) so there should be also `model.query()`-api available. This is important because find doesn't support features we need from TMI API..

Comment (olkorhon) [inline app/controllers/index.js]: Thought it was a deprecated call or something similar since I couldn't find references to it in the mongoose Model documentation. Seems to work though, reverting..

File: app/models/testcase.js [hunk 2]
```diff
@@ -240,7 +240,9 @@ TestCaseSchema.static({
  * Register
  */
 TestCaseSchema.plugin( QueryPlugin ); //install QueryPlugin
+
 var Testcase = mongoose.model('Testcase', TestCaseSchema);
-Testcase.ensureIndexes(function (err) {
-  if (err) return handleError(err);
-});
+// TODO figure out how to synchronously ensureIndexes so we do not close db connection before the process is completed
```
Comment (jupe) [inline app/models/testcase.js]: its still better to be asynchronous. Solution could be that we emit some common event bus message that indexes is ensured before allow clients to be connected to API's...  -> "kind of boot sequence"...

File: app/controllers/resources.js [hunk 3]
```diff
@@ -72,26 +77,25 @@ class ResourcesController extends DefaultController {
   static allocMultiple(req, res) {
     req.Resource.allocateResources(req.body, (error, allocated) => {
       if (error) {
-        res.status(404).json(error);
+        res.status(404).json({ error });
```
Comment (jupe) [inline app/controllers/resources.js]: why this change ?

File: app/controllers/authentication.js [hunk 4]
```diff
@@ -1,229 +1,238 @@
...
-            callback({ status: 500, msg: err.toString() });
-          } else {
-            var belongsOrg = _.find(response.body, { login: nconf.get('github').organization });
+    const checkOrganization = (accessToken, headers, profile, callback) => {
+      request.get({ url: `${userApiUrl}/orgs`, qs: accessToken, headers, json: true }, (err, response) => {
+        if (err) {
+          winston.info('Github auth: checkOrganization error');
+          callback({ status: 500, msg: err.toString() });
+        } else {
+          const belongsOrg = _.find(response.body, { login: nconf.get('github').organization });
```
Comment (jupe) [inline app/controllers/authentication.js]: move `nconf.get..organization` to top of the file as const.

+
add env.example.json:
```
  "GITHUB_ORG": "ORGANIZATION",
  "GITHUB_ADMINTEAM": "ADMIN-TEAM",
```

File: app/controllers/authentication.js [hunk 5]
```diff
@@ -1,229 +1,238 @@
...
-            if (belongsOrg) {
-              callback(null, accessToken, headers, profile);
-            } else {
-              winston.info('Github auth: user not in organization');
-              callback({ status: 401, msg: 'You do not have required access.' });
-            }
+          if (belongsOrg) {
+            callback(null, accessToken, headers, profile);
+          } else {
+            winston.info('Github auth: user not in organization');
```
Comment (jupe) [inline app/controllers/authentication.js]: use warning level + include admin team name to log

File: app/controllers/authentication.js [hunk 6]
```diff
@@ -1,229 +1,238 @@
...
-            });
-            profile.group = isAdmin ? 'admins' : '';
+    const checkAdmin = (accessToken, headers, profile, callback) => {
+      request.get({ url: `${userApiUrl}/teams`, qs: accessToken, headers, json: true }, (err, response) => {
+        if (err) {
+          winston.info('Github auth: checkAdmin error');
+          callback({ status: 500, msg: err.toString() });
+        } else {
+          const isAdmin = _.find(response.body, (team) => {
+            return team.name === nconf.get('github').adminTeam && team.organization.login === nconf.get('github').organization;
```
Comment (jupe) [inline app/controllers/authentication.js]: move `nconf.get..`:s to top of file as const

File: app/controllers/authentication.js [hunk 7]
```diff
@@ -1,229 +1,238 @@
...
-        function (err, response, accessToken) {
-          if (err) {
-            winston.info('Github auth: authorization error');
-            callback({ status: 500, msg: err.toString() });
-          } else {
-            accessToken = qs.parse(accessToken);
-            var headers = { 'User-Agent': 'Satellizer' };
+      request.get({ url: accessTokenUrl, qs: params }, (err, response, accessToken) => {
+        if (err) {
+          winston.info('Github auth: authorization error');
```
Comment (jupe) [inline app/controllers/authentication.js]: use warning level

File: app/controllers/authentication.js [hunk 8]
```diff
@@ -1,229 +1,238 @@
...
-            winston.info('Github auth: no email error');
-            callback({ status: 409, msg: 'Did not get email address.' });
-          } else {
-            callback(null, accessToken, headers, profile);
-          }
-        });
+    const getProfile = (accessToken, headers, callback) => {
+      request.get({ url: userApiUrl, qs: accessToken, headers, json: true }, (err, response, profile) => {
+        if (err) {
+          winston.info('Github auth: getProfile error');
```
Comment (jupe) [inline app/controllers/authentication.js]: use warn level

File: app/controllers/authentication.js [hunk 9]
```diff
@@ -1,229 +1,238 @@
...
+          callback({ status: 500, msg: err.toString() });
+        } else if (!err && response.statusCode !== 200) {
+          winston.info('Github auth: bad profile response');
```
Comment (jupe) [inline app/controllers/authentication.js]: warn/debug level

File: app/controllers/authentication.js [hunk 10]
```diff
@@ -1,229 +1,238 @@
...
+          callback({ status: 409, msg: 'Did not get github profile.' });
+        } else if (!profile.email) {
+          winston.info('Github auth: no email error');
```
Comment (jupe) [inline app/controllers/authentication.js]: warn level. This is by the way problematic: GitHub default configuration use email as private so its not visible here. We have to change email as optional field probably..

File: app/controllers/authentication.js [hunk 11]
```diff
@@ -1,229 +1,238 @@
...
-        function (err, response) {
-          if (err) {
-            winston.info('Github auth: checkOrganization error');
-            callback({ status: 500, msg: err.toString() });
-          } else {
-            var belongsOrg = _.find(response.body, { login: nconf.get('github').organization });
+    const checkOrganization = (accessToken, headers, profile, callback) => {
+      request.get({ url: `${userApiUrl}/orgs`, qs: accessToken, headers, json: true }, (err, response) => {
+        if (err) {
+          winston.info('Github auth: checkOrganization error');
```
Comment (jupe) [inline app/controllers/authentication.js]: warn level

File: app/controllers/testcases.js [hunk 12]
```diff
@@ -3,40 +3,24 @@
...
+  download(req, res) {
+    const tc = req.testcase.toObject();
     if (tc.files.length > 0) {
       res.attachment();
-      var url = tc.files[0].href;
+      const url = tc.files[0].href;
       request({
         followAllRedirects: true,
-        url: url,
+        url,
```
Comment (olkorhon) [inline app/controllers/testcases.js]: Eslint by default prefers new Object notation where variables can be initialized to objects with their own name as key, I still agree that mixing the two notation methods in single object looks weird.

See: - new notations in ECMAScript 2015 -
https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Object_initializer

File: app/controllers/authentication.js (hunk 10 above)
Comment (olkorhon) [inline app/controllers/authentication.js]: Good point but we still need some user-friendly way to search for users(before it was email). Github username could work, but in any case, one or more fields of user needs to be searchable.
